core/src/main/java/org/elasticsearch/action/resync/TransportResyncReplicationAction.java

```diff
@@ -0,0 +1,158 @@
+import org.elasticsearch.indices.IndicesService;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportException;
+import org.elasticsearch.transport.TransportResponseHandler;
+import org.elasticsearch.transport.TransportService;
+
+import java.util.function.Supplier;
+
+public class TransportResyncReplicationAction extends TransportWriteAction<ResyncReplicationRequest,
+    ResyncReplicationRequest, ResyncReplicationResponse> implements PrimaryReplicaSyncer.SyncAction {
+
+    public static String ACTION_NAME = "indices:admin/seq_no/resync";
+
+    @Inject
+    public TransportResyncReplicationAction(Settings settings, TransportService transportService,
+                                            ClusterService clusterService, IndicesService indicesService, ThreadPool threadPool,
+                                            ShardStateAction shardStateAction, ActionFilters actionFilters,
+                                            IndexNameExpressionResolver indexNameExpressionResolver) {
+        super(settings, ACTION_NAME, transportService, clusterService, indicesService, threadPool, shardStateAction, actionFilters,
+            indexNameExpressionResolver, ResyncReplicationRequest::new, ResyncReplicationRequest::new, ThreadPool.Names.BULK);
+    }
+
+    @Override
+    protected void registerRequestHandlers(String actionName, TransportService transportService, Supplier<ResyncReplicationRequest> request,
+                                           Supplier<ResyncReplicationRequest> replicaRequest, String executor) {
+        transportService.registerRequestHandler(actionName, request, ThreadPool.Names.SAME, new OperationTransportHandler());
+        // we should never reject resync because of thread pool capacity on primary
+        transportService.registerRequestHandler(transportPrimaryAction,
+            () -> new ConcreteShardRequest<>(request),
+            executor, true, true,
+            new PrimaryOperationTransportHandler());
+        transportService.registerRequestHandler(transportReplicaAction,
+            () -> new ConcreteReplicaRequest<>(replicaRequest),
+            executor, true, true,
+            new ReplicaOperationTransportHandler());
+    }
+
+    @Override
+    protected ResyncReplicationResponse newResponseInstance() {
+        return new ResyncReplicationResponse();
+    }
+
+    @Override
+    protected ReplicationOperation.Replicas newReplicasProxy() {
+        // We treat the resync as best-effort for now and don't mark unavailable shard copies as stale.
+        return new ReplicasProxy();
+    }
+
+    @Override
+    protected WritePrimaryResult<ResyncReplicationRequest, ResyncReplicationResponse> shardOperationOnPrimary(
+        ResyncReplicationRequest request, IndexShard primary) throws Exception {
+        final ResyncReplicationRequest replicaRequest = performOnPrimary(request, primary);
```

Comment:
It's used in the tests. This ensures that the ResyncAction in the tests stays in sync with TransportResyncReplicationAction (e.g. if we would add more code there).